Open Judge System/Workers/OJS.Workers.LocalWorker/SubmissionJob.cs

```diff
@@ -46,39 +47,45 @@ public void Start()
             {
                 var data = new OjsData();
                 Submission submission;
-                try
-                {
-                    submission = data.Submissions.GetSubmissionForProcessing();
-                }
-                catch (Exception exception)
-                {
-                    this.logger.FatalFormat("Unable to get submission for processing. Exception: {0}", exception);
-                    throw;
-                }
-
-                if (submission == null)
-                {
-                    // No submission available. Wait 1 second and try again.
-                    Thread.Sleep(1000);
-                    continue;
-                }
-
-                if (!this.processingSubmissionIds.Add(submission.Id))
-                {
-                    // Other thread is processing the same submission. Wait the other thread to set Processing to true and then try again.
-                    Thread.Sleep(100);
-                    continue;
-                }
+                int submissionId;
 
                 try
                 {
-                    submission.Processing = true;
-                    data.SaveChanges();
+                    var retrievedSubmissionSuccessfully = false;
+                    lock (this.submissionsForProcessing)
+                    {
+                        if (this.submissionsForProcessing.IsEmpty)
+                        {
+                            var submissions = data
+                                .Submissions
+                                .All()
+                                .Where(x => !x.Processed && !x.Processing)
+                                .OrderBy(x => x.Id)
+                                .Select(x => x.Id)
+                                .ToList();
+
+                            submissions.ForEach(this.submissionsForProcessing.Enqueue);
+                        }
+
+                        retrievedSubmissionSuccessfully = this.submissionsForProcessing.TryDequeue(out submissionId);                     
+                    }
+
+                    if (retrievedSubmissionSuccessfully)
+                    {
+                        this.logger.InfoFormat("Submission №{0} retrieved from database successfully", submissionId);
```

Comment:
`retrieved from database` is not always correct. Maybe you can remove `from database`